add Network.config.openshift.io validation

danwinship opened the pull request:
Use the new CRD validator hack to validate Network.config.openshift.io.

Fixes the half of https://github.com/openshift/cluster-network-operator/issues/69 that wasn't fixed by https://github.com/openshift/cluster-network-operator/pull/76.

danwinship commented:
/retest

enj commented on pkg/admission/customresourcevalidation/network/validate_network.go:
```diff
@@ -0,0 +1,130 @@
+	allErrs := field.ErrorList{}
+
+	if len(spec.ClusterNetwork) == 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterNetwork"), nil, "must have at least 1 entry"))
```
This is a required error, not invalid.

enj commented on pkg/admission/customresourcevalidation/network/validate_network.go:
```diff
@@ -0,0 +1,130 @@
+		}
+	}
+
+	if spec.NetworkType == "" {
```
This is an enum right?  Validate as such.  Also we really should update to match the naming and structure of the Authentication type field.

danwinship replied:
It is an enum, but we will be adding new elements to it over time, hopefully without needing other changes to this type. (This is just for "basic" config; "complicated" config happens in the network operator's config type.)

We (@openshift/sig-networking) probably need to think about the details of how this is going to work some more... Maybe it really only makes sense to allow `"OpenShiftSDN"`, `"OVNKubernetes"`, and `"Other"` in the config.openshift.io object, since if you're using a non-openshift-default networking type, it's pretty much guaranteed that you'll need to specify additional configuration options in the network operator config anyway...

enj replied:
Consider `Authentication` as an example (and note that the name of the json field is `type` and not `authenticationType`:

```go
type AuthenticationSpec struct {
	// type identifies the cluster managed, user facing authentication mode in use.
	// Specifically, it manages the component that responds to login attempts.
	// The default is IntegratedOAuth.
	Type AuthenticationType `json:"type"`
...
```

```go
type AuthenticationType string

const (
	// None means that no cluster managed authentication system is in place.
	// Note that user login will only work if a manually configured system is in place and
	// referenced in authentication spec via oauthMetadata and webhookTokenAuthenticators.
	AuthenticationTypeNone AuthenticationType = "None"

	// IntegratedOAuth refers to the cluster managed OAuth server.
	// It is configured via the top level OAuth config.
	AuthenticationTypeIntegratedOAuth AuthenticationType = "IntegratedOAuth"

	// TODO if we add support for an in-cluster operator managed Keycloak instance
	// AuthenticationTypeKeycloak AuthenticationType = "Keycloak"
)
```

```go
// OAuth holds cluster-wide information about OAuth.  The canonical name is `cluster`.
// It is used to configure the integrated OAuth server.
// This configuration is only honored when the top level Authentication config has type set to IntegratedOAuth.
type OAuth struct {
... 
}
```

The only configuration we handle through the operator is the standard `OperatorSpec` and `GenericOperatorConfig`.

enj commented on pkg/admission/customresourcevalidation/network/validate_network.go:
```diff
@@ -0,0 +1,130 @@
+		}
+	}
+
+	if len(spec.ServiceNetwork) != 1 {
```
If there can only be one, why is this a list?

List aside, checking for 0 and >1 separately will allow for better error messages.

danwinship replied:
Eventually we will support multiple CIDRs (eg, one IPv4 and one IPv6 network, but there are other use cases too). But k8s 1.12 only supports a single CIDR so we can reject anything else for now.

danwinship commented:
- Changed 0-length `spec.clusterNetwork` and `spec.serviceNetwork` to `field.Required` rather than `field.Invalid`.
- Split out a separate `field.Invalid` for multi-valued `spec.serviceNetwork`
- Dropped `detail` argument from the field errors where they were redundant

I didn't change `NetworkType` to `Type` because that would have to happen in api and then get propagated here (and the installer and cluster-network-operator) but I can do that if you think the naming consistency is important...

danwinship commented:
/retest

danwinship commented:
/retest

danwinship commented:
/retest

danwinship commented:
/retest

danwinship commented:
/remove-lifecycle rotten
/lifecycle frozen

danwinship commented:
/retest

squeed commented:
I think the right way to install `config.openshift.io` validation is by adding kubebulder magic comments to the structs - https://book.kubebuilder.io/reference/generating-crd.html#validation.

It's implemented by the cluster config operator - it automatically generates the CRDs from the go types.

danwinship commented:
> I think the right way to install `config.openshift.io` validation is by adding kubebulder magic comments to the structs

Right. Filed https://github.com/openshift/api/pull/412
